dan_layer/storage_sqlite/src/sqlite_shard_store_factory.rs

```diff
@@ -610,28 +651,87 @@ impl ShardStoreTransaction<PublicKey, TariDanPayload> for SqliteShardStoreTransa
     fn save_substate_changes(
         &mut self,
         changes: HashMap<ShardId, Option<SubstateState>>,
-        node: TreeNodeHash,
+        node: &HotStuffTreeNode<PublicKey>,
     ) -> Result<(), Self::Error> {
-        for (sid, st_ch) in &changes {
+        use crate::schema::substates::{data, is_draft, justify, node_height, shard_id, substate_type, tree_node_hash};
+        let payload_id = Vec::from(node.payload().as_slice());
+        for (sid, st_change) in &changes {
             let shard = Vec::from(sid.as_bytes());
-            let substate_change = if let Some(st_ch) = st_ch {
-                json!(st_ch).to_string()
-            } else {
-                "".to_string()
-            };
-
-            let new_row = NewSubStateChange {
-                shard_id: shard,
-                tree_node_hash: Vec::from(node.as_bytes()),
-                substate_change,
-            };
-
-            diesel::insert_into(substate_changes)
-                .values(&new_row)
-                .execute(&self.connection)
-                .map_err(|e| Self::Error::QueryError {
-                    reason: format!("Save substate change: {}", e),
-                })?;
+
+            let rows_affected = diesel::update(
```

Comment:
Looks like upserts are a Diesel 2.0 feature. A raw query is probably easiet `REPLACE OR INSERT INTO` 
and would greatly shorten this code.